typed-store/src/rocks/errors.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright(C) 2022, Mysten Labs
+// SPDX-License-Identifier: Apache-2.0
+
+use bincode::ErrorKind as BincodeErrorKind;
+
+use rocksdb::Error as RocksError;
+use serde::{Deserialize, Serialize};
+use std::{fmt, fmt::Display};
+use thiserror::Error;
+
+#[non_exhaustive]
+#[derive(Error, Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone)]
+pub enum TypedStoreError {
+    #[error("rocksdb error")]
+    RocksDBError(#[from] RocksErrorDef),
+    #[error("(de)serialization error")]
+    SerializationError(#[from] BincodeErrorDef),
+    #[error("the column family {0} was not registered with the database")]
+    UnregisteredColumn(String),
+    #[error("a batch operation can't operate across databases")]
+    CrossDBBatch,
+}
+
+#[derive(Serialize, Deserialize, Clone, Eq, PartialEq, Hash, Debug, Error)]
+pub struct RocksErrorDef {
+    message: String,
+}
+
+impl From<RocksError> for RocksErrorDef {
+    fn from(err: RocksError) -> Self {
+        RocksErrorDef {
+            message: err.as_ref().to_string(),
+        }
+    }
+}
+
+impl From<RocksError> for TypedStoreError {
+    fn from(err: RocksError) -> Self {
+        TypedStoreError::RocksDBError(err.into())
+    }
+}
+
+impl Display for RocksErrorDef {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        self.message.fmt(formatter)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Hash, Eq, PartialEq, Debug, Error)]
+pub enum BincodeErrorDef {
```

Comment:
Here are we tying a little too closely the error reporting with the underlying choice of using bincode? Maybe we could compress some of these cases, unless we believe devs using this lib can take a different action depending on the nature of the serialization/deserialization error?

Reply:
Yes, indeed. However, the classical technique to do that (a severly constrained trait object) is hampered by us requiring fancy traits (`Clone, Hash, Eq`) on this type because we put the `Result` directly in `HashMap`s.

in the meantine I've found a worse solution around it, but that's probably something we want to fix on the logn term.